### PR TITLE
Avoid capturing this unsafely in MultiFrameCodec

### PR DESCRIPTION
**Pull request by dnfield**
fixes flutter/flutter#13359

Creates a `State` object to share as a `shared_ptr` instead of capturing this. I don't believe we can use a `fml::WeakPtr`, since we would have to create it on the IO task runner, which might be blocked until after the MultiFrameCodec is destroyed on the UI runner.

Avoids decoding the image if the collect happens before we start image decoding - this seems like the right thing to do to me, but in honesty if I don't do this the test is segfaulting when I try to drain the unref queue and I'm not quite sure why.

Moves some of the logic that's sharable from the Dart Isolate unit tests to a common file for use by the image decoder tests.

### CODE REVIEW

**Comment by Qxyat** [comment]
We encounter a problem which is the app crash at the time the engine exit due to `MultiFrameCodec` share between `ui` and `io` thread. When this one will land?

**Comment by dnfield** [comment]
Ahh this was passing on macOS but not Linux.

I'll fix up the test harness and then it just needs a review.

File: lib/ui/painting/image_decoder_unittests.cc [hunk 1]
```diff
@@ -557,5 +612,87 @@ TEST(ImageDecoderTest, VerifySubpixelDecodingPreservesExifOrientation) {
   assert_image(decode({}, 100));
 }
 
+TEST_F(ImageDecoderFixtureTest,
```
Comment (gaaclarke) [inline lib/ui/painting/image_decoder_unittests.cc]: This test is pretty big, it can't be groked with a passing perusal.  I recommend adding a docstring at the beginning of the test, or pulling out some helper functions so that the gist can be parsed quicker.

File: lib/ui/painting/image_decoder_unittests.cc [hunk 2]
```diff
@@ -557,5 +612,87 @@ TEST(ImageDecoderTest, VerifySubpixelDecodingPreservesExifOrientation) {
...
+    EXPECT_FALSE(codec);
+
+    io_latch.Signal();
+
+    latch.Signal();
+  });
+  latch.Wait();
+
+  // Destroy the IO manager
+  runners.GetIOTaskRunner()->PostTask([&]() {
```
Comment (gaaclarke) [inline lib/ui/painting/image_decoder_unittests.cc]: There is alot of work here to serialize async operations with latches.  It would be easier to follow if you made a PerformSyncTask function that takes in a task runner and a lambda and doesn't complete until the task runner finishes the lambda.

I've tried to add this in the past and have met resistance, maybe for tests at least people will be accepting.

File: lib/ui/painting/multi_frame_codec.h [hunk 3]
```diff
@@ -26,24 +26,31 @@ class MultiFrameCodec : public Codec {
...
-
-  sk_sp<SkImage> GetNextFrameImage(fml::WeakPtr<GrContext> resourceContext);
-
-  void GetNextFrameAndInvokeCallback(
-      std::unique_ptr<DartPersistentValue> callback,
-      fml::RefPtr<fml::TaskRunner> ui_task_runner,
-      fml::WeakPtr<GrContext> resourceContext,
-      fml::RefPtr<flutter::SkiaUnrefQueue> unref_queue,
-      size_t trace_id);
+  struct State {
```
Comment (gaaclarke) [inline lib/ui/painting/multi_frame_codec.h]: class docstring

**Comment by dnfield** [comment]
Test harness was missing a config to export the needed symbols on Linux - done and should all be passing now.

File: testing/dart_isolate_runner.h [hunk 4]
```diff
@@ -0,0 +1,198 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
```
Comment (gaaclarke) [inline testing/dart_isolate_runner.h]: This appears to be a a big refactor.  It would be easier to review and manage if this was split into a different PR.

File: lib/ui/painting/image_decoder_unittests.cc (hunk 2 above)
Comment (dnfield) [reply lib/ui/painting/image_decoder_unittests.cc]: This is idiomatic code for how the rest of the engine repo does it. I'd be hesitant to add a new test-only class for this that itself needs to be tested...

File: testing/dart_isolate_runner.h (hunk 4 above)
Comment (dnfield) [reply testing/dart_isolate_runner.h]: I'll split this into a separate PR.

Comment (dnfield) [reply testing/dart_isolate_runner.h]: https://github.com/flutter/engine/pull/16933

File: lib/ui/painting/image_decoder_unittests.cc (hunk 2 above)
Comment (gaaclarke) [reply lib/ui/painting/image_decoder_unittests.cc]: Boilerplate that detracts from the point shouldn't be encouraged.  I wasn't think a new class I was thinking this:
```
void PerformSyncTask(TaskRunner runner, Function func) {
  Latch latch;
  runner->PerformTask({
    func();
    latch.signal();
  });
  latch.wait();
}
```

File: lib/ui/painting/multi_frame_codec.h (hunk 3 above)
Comment (dnfield) [reply lib/ui/painting/multi_frame_codec.h]: Done. It's probably not in proper doxygen format, but this is a private class anyway.

File: lib/ui/painting/image_decoder_unittests.cc (hunk 2 above)
Comment (dnfield) [reply lib/ui/painting/image_decoder_unittests.cc]: The synchronization work here is also highly constructed to simulate collecting the engine while IO work is ongoing in a repeatable, non-flaky way.  I'm not sure it would help to have something like `PerformSyncTask`, since I need to specifically interleave certain actions.  I hope the comment I added at the top clarifies.

File: lib/ui/painting/image_decoder_unittests.cc (hunk 1 above)
Comment (dnfield) [reply lib/ui/painting/image_decoder_unittests.cc]: Added a comment to the top.  I'm not quite sure how to refactor this into smaller functions while still preserving the ordering that's needed, but maybe when I get in we can talk about it some more offline.

File: lib/ui/painting/multi_frame_codec.cc [hunk 5]
```diff
@@ -164,12 +177,13 @@ Dart_Handle MultiFrameCodec::getNextFrame(Dart_Handle callback_handle) {
 
   const auto& task_runners = dart_state->GetTaskRunners();
 
-  task_runners.GetIOTaskRunner()->PostTask(fml::MakeCopyable(
-      [callback = std::make_unique<DartPersistentValue>(
-           tonic::DartState::Current(), callback_handle),
-       this, trace_id, ui_task_runner = task_runners.GetUITaskRunner(),
-       io_manager = dart_state->GetIOManager()]() mutable {
-        GetNextFrameAndInvokeCallback(
```
[the remark this answers is not in the document]
Comment (dnfield) [reply lib/ui/painting/multi_frame_codec.cc]: We can't use a WeakPtr here because we have to access the object across threads, and WeakPtr is not thread safe.

[the remark this answers is not in the document]
Comment (dnfield) [reply lib/ui/painting/multi_frame_codec.cc]: And we can't create the weak ptr on the IO task runner because our test case would fail - e.g. if the IO runner is blocked, and the object gets collected before the task to create the weak ptr  runs

File: lib/ui/painting/image_decoder_unittests.cc (hunk 2 above)
Comment (dnfield) [reply lib/ui/painting/image_decoder_unittests.cc]: On this specific test, I already have the latch and would need to have it anyway.  The added function would only save a line or two of code here :\

Comment (gaaclarke) [reply lib/ui/painting/image_decoder_unittests.cc]: The same thing could be said of for-loops versus foreach-loops.  It isn't about LOC, its about avoiding errors, extra variables, and having a meaningful abstraction with which to build off of.

Comment (dnfield) [reply lib/ui/painting/image_decoder_unittests.cc]: Yes - and I think such a decision on style is something we should follow consistently in the repo. 

I'm not opposed to making this change so much as opposed to doing it differently just here, or even just in this test file, compared to the way we do it everywhere else. It'd be easier to just refactor all the callsites of this if they're all pretty similar, rather than doing it specially here.

File: lib/ui/painting/multi_frame_codec.cc [hunk 6]
```diff
@@ -164,12 +177,13 @@ Dart_Handle MultiFrameCodec::getNextFrame(Dart_Handle callback_handle) {
...
+  task_runners.GetIOTaskRunner()->PostTask(
+      fml::MakeCopyable([callback = std::make_unique<DartPersistentValue>(
+                             tonic::DartState::Current(), callback_handle),
+                         state = state_, trace_id,
+                         ui_task_runner = task_runners.GetUITaskRunner(),
+                         io_manager = dart_state->GetIOManager()]() mutable {
+        state->GetNextFrameAndInvokeCallback(
```
[the remark this answers is not in the document]
Comment (chinmaygarde) [reply lib/ui/painting/multi_frame_codec.cc]: > you can do what you are doing in line 144.

Wait. Why are we even invoking the next frame callback on a muli-frame codec that has died (which would be indicated by the weak pointer lock failing)?

Instead of invoking the next frame callback when the weak pointer lock fails, scheduled a task onto the UI thread that just releases the callback but not invoke it (that has Dart state and you cant just drop it on the IO thread).

File: lib/ui/painting/image_decoder_unittests.cc [hunk 7]
```diff
@@ -557,5 +612,100 @@ TEST(ImageDecoderTest, VerifySubpixelDecodingPreservesExifOrientation) {
...
+  //   - Starts a Dart VM
+  //   - Latches the IO task runner
+  //   - Create a MultiFrameCodec for an animated gif pointed to a callback
+  //     in the Dart fixture
+  //   - Calls getNextFrame on the UI task runner
+  //   - Collects the MultiFrameCodec object before unlatching the IO task
+  //     runner.
+  //   - Unlatches the IO task runner
+  auto settings = CreateSettingsForFixture();
+  settings.leak_vm = false;
```
[the remark this answers is not in the document]
Comment (dnfield) [reply lib/ui/painting/image_decoder_unittests.cc]: done

File: lib/ui/painting/image_decoder_unittests.cc [hunk 8]
```diff
@@ -14,7 +20,56 @@
...
+      return;
+    }
+
+    settings.assets_dir = assets_dir_.get();
+
+    // In JIT execution, all snapshots are present within the binary itself and
+    // don't need to be explicitly supplied by the embedder. In AOT, these
+    // snapshots will be present in the application AOT dylib.
+    if (DartVM::IsRunningPrecompiledCode()) {
+      PrepareSettingsForAOTWithSymbols(settings, aot_symbols_);
```
[the remark this answers is not in the document]
Comment (dnfield) [reply lib/ui/painting/image_decoder_unittests.cc]: done

File: lib/ui/painting/image_decoder_unittests.cc [hunk 9]
```diff
@@ -557,5 +612,100 @@ TEST(ImageDecoderTest, VerifySubpixelDecodingPreservesExifOrientation) {
...
+  auto vm_data = vm_ref.GetVMData();
+
+  auto gif_mapping = OpenFixtureAsSkData("hello_loop_2.gif");
+
+  ASSERT_TRUE(gif_mapping);
+
+  auto gif_codec = SkCodec::MakeFromData(gif_mapping);
+  ASSERT_TRUE(gif_codec);
+
+  auto loop = fml::ConcurrentMessageLoop::Create();
```
[the remark this answers is not in the document]
Comment (dnfield) [reply lib/ui/painting/image_decoder_unittests.cc]: Removed - this was copypasted from another test and I missed dropping it during refactoring.

File: lib/ui/painting/image_decoder_unittests.cc [hunk 10]
```diff
@@ -557,5 +612,100 @@ TEST(ImageDecoderTest, VerifySubpixelDecodingPreservesExifOrientation) {
...
+
+  // Latch the IO task runner.
+  runners.GetIOTaskRunner()->PostTask([&]() { io_latch.Wait(); });
+
+  runners.GetUITaskRunner()->PostTask([&]() {
+    fml::AutoResetWaitableEvent isolate_latch;
+    fml::RefPtr<MultiFrameCodec> codec;
+    EXPECT_TRUE(isolate->RunInIsolateScope([&]() -> bool {
+      Dart_Handle library = Dart_LookupLibrary(Dart_NewStringFromCString(
+          "package:engine_root/ui/fixtures/ui_test.dart"));
```
[the remark this answers is not in the document]
Comment (dnfield) [reply lib/ui/painting/image_decoder_unittests.cc]: It does, I didn't know about `Dart_RootLibrary`.  Done.

File: lib/ui/painting/multi_frame_codec.h [hunk 11]
```diff
@@ -26,24 +26,44 @@ class MultiFrameCodec : public Codec {
...
+  // Instead, the MultiFrameCodec creates this object when it is constructed,
+  // shares it with the IO task runner's decoding work, and sets the live_
+  // member to false when it is destructed.
+  struct State {
+    State(std::unique_ptr<SkCodec> codec);
+
+    const std::unique_ptr<SkCodec> codec_;
+    const int frameCount_;
+    const int repetitionCount_;
+    int nextFrameIndex_;
```
[the remark this answers is not in the document]
Comment (dnfield) [reply lib/ui/painting/multi_frame_codec.h]: Done.  I'd like to avoid the mutex since this is a potentially performance critical path, e.g. for an animated image firing on every vsync.

File: lib/ui/painting/multi_frame_codec.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (dnfield) [reply lib/ui/painting/multi_frame_codec.cc]: I'm not clear on where it's safe to create the weak_ptr for the State.  I can't do it on the UI thread, and by the time I get to the IO thread the UI object might have been collected.

[the remark this answers is not in the document]
Comment (dnfield) [reply lib/ui/painting/multi_frame_codec.cc]: Ohhh you mean a std::weak_ptr.  I get it now.

[the remark this answers is not in the document]
Comment (dnfield) [reply lib/ui/painting/multi_frame_codec.cc]: For some reason I was thinking we had to still make the callback because some Dart code might be `await`ing the completion of the callback, but now that I think about it that doesn't make sense, since the only way this can happen is if the engine is collected.  If anyone is actually awaiting the callback, the codec must still be alive.  I'm correct in that thinking, right?

[the remark this answers is not in the document]
Comment (dnfield) [reply lib/ui/painting/multi_frame_codec.cc]: Done - I believe

[the remark this answers is not in the document]
Comment (dnfield) [reply lib/ui/painting/multi_frame_codec.cc]: Only question is whether we need to guard a bit more so that if we get collected _after_ this (e.g. we're still alive before we tell Skia to decode a frame, but we're dead after Skia finishes) - will this still be safe?

File: lib/ui/painting/image_decoder_unittests.cc [hunk 12]
```diff
@@ -14,7 +20,56 @@
 namespace flutter {
 namespace testing {
 
-using ImageDecoderFixtureTest = ThreadTest;
+class ImageDecoderFixtureTest : public ThreadTest {
```
[the remark this answers is not in the document]
Comment (dnfield) [reply lib/ui/painting/image_decoder_unittests.cc]: Done

File: lib/ui/painting/multi_frame_codec.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (chinmaygarde) [reply lib/ui/painting/multi_frame_codec.cc]: The lock will fail in that case.

**Comment by dnfield** [comment]
I've run this test with the same random seed that flaked over 100 times on both mac and linux locally in profile mode - it hasn't failed.  I'm retrying the bot here: https://ci.chromium.org/p/flutter/builders/try/Linux%20Host%20Engine/4950.  If it succeeds I plan to land it.

The mac web failure is a flake that has nothing to do with this PR.

**Comment by dnfield** [comment]
FWIW, the flake on Linux host_profile didn't even happen in the new test or codepath touched by this PR :\

**Comment by dnfield** [comment]
Rerun passed.  I'll monitor post submit for failures and revert if it does fail again.
